editor/editor_file_system.cpp

```diff
@@ -115,6 +115,17 @@ String EditorFileSystemDirectory::get_file_path(int p_idx) const {
 	return "res://" + file;
 }
 
+String EditorFileSystemDirectory::get_file_attached_script(int p_idx) const {
+	Vector<String> deps = get_file_deps(p_idx);
+	for (int i = 0; i < deps.size(); i++) {
+		if (deps[i].ends_with(".gd") || deps[i].ends_with(".cs")) {
```

Comment:
Rather than checking for specific languages' extensions, it's better to use `ResourceLoader::get_resource_type(p_path)` to extract a typename and then use an inheritance check to see if it inherits from the `Script` class. If so, you know it's a script, regardless of what language it is.

You can *probably* assume that a user-defined type WON'T be an implementation of a Script, so you could skip it, but ordinarily, you would first check `ScriptServer::is_global_class()` and then based on that either use `EditorNode::get_editor_data().script_class_is_parent()` if true or `ClassDB::is_parent_class()` if false.

Reply:
I shouldn't have to worry about this running outside of the editor right? If so, then is there a way I can check if I'm in the editor?

Reply:
@Atlinx Also, I just realized examining this, that this is not really a good solution (or at least, it doesn't seem very reliable). Any given file can have 0 or more dependencies, any number of which might happen to be a script. But you don't know *which* of those dependencies is the `script` property of the file in question. So, imo, you really *do* need to have some sort of resource script type property being tracked in the `FileInfo`/`FileCache` in `EditroFileSystem`. And rather than forcing the use of `ResourceFormatLoaderText`, you would need to define a virtual function in `ResourceFormatLoader` that gets the scripted type of the resource, if it has one, which you would then override in the various derived implementations. Then the logic to get the file's resource script type would involve iterating over all of the loaders present in the `ResourceLoader`.

Reply:
Does this mean we would have to break compatibility with 3.x? Adding another FileInfo means incrementing the file cache file version, and I'm not sure if this counts as compatibility breakage. Either way, I'll try to rewrite this PR using your suggestions.

Reply:
I'm *guessing* it would involve breaking compatibility. Even if you assume that it's possible for 3.x and 4.x to have their own branching continuations of file cache versions, the fact that they share the same sequential numeric versioning scheme means that as soon as 3.x attempts to increment beyond where it already is, it will be stepping into the realm of 4.x's prior iterations (probably?). So yeah, I doubt this is something that can be properly done in 3.x, simply because the EditorFileSystem doesn't have any mechanism by which it can efficiently keep track of the type of the script that may or may not be assigned to a given resource file.

Reply:
The ResourceFormatLoader thing might also become a headache *IF* you end up also modifying `get_resource_type` to return a possible non-C++ class as there are many calls to this method throughout the codebase, all of which likely expect it to return an in-engine C++ type rather than a potential scripted type. If you don't change the logic of that method though, and instead define a new function to access the extra information, thereby demanding that any scenario that need it specifically call that separate function, then there's likely to be less conflict with other uses.